### ISSUE TITLE
no-array-index-key not working in React.Children.map

### ISSUE THREAD

**Issue by himynameisdave** [yannickcr/eslint-plugin-react#2083]
Hi! This currently does not flag an error with `no-array-index-key`:

```jsx
React.Children.map(this.props.children, (child, index) => {
    const className = classnames(
                        child.props.className,
                        `${this.props.transition}-finished`
                    );
    return React.cloneElement(child, { key: index, className });
});
```

Specifically, I would expect the `key: index` part to trigger the `no-array-index-key` warning.

(I realize that there is probably a better way to write this (ie: not using `cloneElement`), but I noticed this recently and saw that it didn't flag the warning.

---

PS: Thanks for the awesome ESLint plugin! 🎉  Can't wait for the next release!

![mcbook](https://user-images.githubusercontent.com/4298089/49838152-1cfe9680-fd5e-11e8-921f-2f1e22ac45f9.gif)

**Comment by ljharb** [comment]
Sure, this seems like a reasonable addition.

**Comment by himynameisdave** [comment]
@ljharb thanks! I would actually like to try fixing this (as I use this plugin a lot, so it would be cool to contribute to it).
